### PR TITLE
service/rds: Initial support for RDS Aurora Global Databases

### LINKED ISSUE #6846: Feature Request: Add engine mode for global database to aws_rds_cluster
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Global databases were added to aurora to make cross region replication easier to implement and manage. https://aws.amazon.com/rds/aurora/global-database/. 

### New or Affected Resource(s)

 
* aws_rds_cluster

### Potential Terraform Configuration

 

```hcl
resource "aws_rds_cluster" "default" {
  cluster_identifier      = "aurora-cluster-demo"
  availability_zones      = ["us-west-2a", "us-west-2b", "us-west-2c"]
  regions = ["us-east-1","us-west-2"]
  database_name           = "mydb"
  master_username         = "foo"
  master_password         = "bar"
  backup_retention_period = 5
  preferred_backup_window = "07:00-09:00"
  engine                  = "aurora"
  engine_mode             = "global"
}

```

### References

* https://www.terraform.io/docs/providers/aws/r/rds_cluster.html#engine_mode
* https://aws.amazon.com/rds/aurora/global-database/
* #0000

### PR DESCRIPTION
**Pull request by bflad**
Closes #6846 

Changes:

* resource/aws_rds_cluster: Allow `global` in `engine_mode` validation
* resource/aws_rds_cluster: Add `global_cluster_identifier` argument
* resource/aws_rds_cluster: Automatically remove cluster from global cluster on delete
* New Resource: `aws_rds_global_cluster`

```
--- PASS: TestAccAWSRDSCluster_BacktrackWindow (185.53s)
--- PASS: TestAccAWSRDSCluster_backupsUpdate (206.64s)
--- PASS: TestAccAWSRDSCluster_basic (135.45s)
--- PASS: TestAccAWSRDSCluster_DeletionProtection (165.47s)
--- PASS: TestAccAWSRDSCluster_encrypted (126.42s)
--- PASS: TestAccAWSRDSCluster_EncryptedCrossRegionReplication (1609.46s)
--- PASS: TestAccAWSRDSCluster_EngineMode (344.98s)
--- PASS: TestAccAWSRDSCluster_EngineMode_Global (118.41s)
--- PASS: TestAccAWSRDSCluster_EngineMode_ParallelQuery (117.81s)
--- PASS: TestAccAWSRDSCluster_EngineVersion (122.85s)
--- PASS: TestAccAWSRDSCluster_EngineVersionWithPrimaryInstance (1022.02s)
--- PASS: TestAccAWSRDSCluster_generatedName (176.05s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier (125.53s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Add (132.58s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Remove (140.14s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Update (124.50s)
--- PASS: TestAccAWSRDSCluster_iamAuth (115.26s)
--- PASS: TestAccAWSRDSCluster_importBasic (116.22s)
--- PASS: TestAccAWSRDSCluster_kmsKey (144.50s)
--- PASS: TestAccAWSRDSCluster_missingUserNameCausesError (2.31s)
--- PASS: TestAccAWSRDSCluster_namePrefix (145.07s)
--- PASS: TestAccAWSRDSCluster_Port (252.80s)
--- PASS: TestAccAWSRDSCluster_s3Restore (1474.31s)
--- PASS: TestAccAWSRDSCluster_ScalingConfiguration (270.01s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier (438.26s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_DeletionProtection (374.71s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EncryptedRestore (331.91s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineMode_ParallelQuery (353.25s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineMode_Provisioned (355.21s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineVersion_Different (358.34s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineVersion_Equal (394.42s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_PreferredBackupWindow (380.13s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_PreferredMaintenanceWindow (334.25s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_Tags (354.72s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_VpcSecurityGroupIds (255.20s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_VpcSecurityGroupIds_Tags (271.42s)
--- PASS: TestAccAWSRDSCluster_takeFinalSnapshot (93.01s)
--- PASS: TestAccAWSRDSCluster_updateCloudwatchLogsExports (158.10s)
--- PASS: TestAccAWSRDSCluster_updateIamRoles (122.85s)
--- PASS: TestAccAWSRDSCluster_updateTags (153.91s)

--- PASS: TestAccAWSRdsGlobalCluster_basic (14.24s)
--- PASS: TestAccAWSRdsGlobalCluster_DatabaseName (24.06s)
--- PASS: TestAccAWSRdsGlobalCluster_DeletionProtection (22.31s)
--- PASS: TestAccAWSRdsGlobalCluster_disappears (11.16s)
--- PASS: TestAccAWSRdsGlobalCluster_Engine_Aurora (14.41s)
--- PASS: TestAccAWSRdsGlobalCluster_EngineVersion_Aurora (14.24s)
--- PASS: TestAccAWSRdsGlobalCluster_StorageEncrypted (25.60s)
```

### CODE REVIEW

File: aws/resource_aws_rds_cluster.go [hunk 1]
```diff
@@ -1084,6 +1110,30 @@ func resourceAwsRDSClusterUpdate(d *schema.ResourceData, meta interface{}) error
...
+	if d.HasChange("global_cluster_identifier") {
+		oRaw, nRaw := d.GetChange("global_cluster_identifier")
+		o := oRaw.(string)
+		n := nRaw.(string)
+
+		if o == "" {
+			return errors.New("Existing RDS Clusters cannot be added to an existing RDS Global Cluster")
+		}
+
+		if o != "" && n != "" {
```
Comment (mbfrahry) [inline aws/resource_aws_rds_cluster.go]: I think `o != ""` is covered in the above block so we should just check that `n != ""`
```suggestion
		if n != "" {
```

File: aws/resource_aws_rds_cluster.go [hunk 2]
```diff
@@ -1128,6 +1178,22 @@ func resourceAwsRDSClusterDelete(d *schema.ResourceData, meta interface{}) error
 	conn := meta.(*AWSClient).rdsconn
 	log.Printf("[DEBUG] Destroying RDS Cluster (%s)", d.Id())
 
+	// Automatically remove from global cluster to bypass this error on deletion:
+	// InvalidDBClusterStateFault: This cluster is a part of a global cluster, please remove it from globalcluster first
+	if d.Get("global_cluster_identifier").(string) != "" {
+		input := &rds.RemoveFromGlobalClusterInput{
```
Comment (mbfrahry) [inline aws/resource_aws_rds_cluster.go]: Since this same chunk of code is used in update, I'd suggest having it in it's own function

File: aws/resource_aws_rds_global_cluster.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+			"global_cluster_identifier": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"global_cluster_resource_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"source_db_cluster_identifier": {
```
Comment (mbfrahry) [inline aws/resource_aws_rds_global_cluster.go]: I don't see this being used anywhere. Should we remove it or was it missed?

File: aws/resource_aws_rds_cluster.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_rds_cluster.go]: The error handling is slightly different. 😄 I'm not sure if the indirection is worthwhile to fill in a two value struct, print a log line, and call the API, but we can certainly discuss this out of band! 👍

File: aws/resource_aws_rds_global_cluster.go (hunk 3 above)
Comment (bflad) [reply aws/resource_aws_rds_global_cluster.go]: Good find and probably something I should've mentioned in the commit message/PR comment!

The RDS API gives you this second option for creating Global Clusters, based on an existing Cluster with global engine mode, instead of creating an "empty" Global Cluster. I think there were two reasons for omitting this to start.

When playing with this, the workflow wound up being very tedious because of operation ordering requirements within the RDS API and Terraform references. If I recall correctly, since the Cluster becomes a child of the Global Cluster in the dependency graph and Global Clusters, it wasn't easy to setup the resources to properly destroy. I may have inadvertently fixed this issue with the `aws_rds_cluster` resource `Delete` function update though. 😅 

The other workflow issue is that engine mode cannot be currently updated for existing Clusters, so essentially you have to create new Clusters anyways to get Global Cluster support. I didn't think we needed to introduce all the complexity of supporting both creation methods initially, since operators can just adjust their initial Global Cluster configuration to match our supported workflow. I may have missed something in my thinking there though.

TL;DR I'll remove this attribute for now since its not used. If there are use cases that require it, we can add it (and all its complexity) later. 👍

Comment (bflad) [reply aws/resource_aws_rds_global_cluster.go]: Now that I'm thinking about it more -- it was that because of the referencing (Global Cluster being dependent on Cluster), we'd try to destroy the Global Cluster first and it would complain that it still had members. We'd have to introduce a flag in `aws_rds_global_cluster` to allow you to force remove all members first. Not unsurmountable, but definitely additional complexity which we might not need.

**Comment by bflad** [comment]
TravisCI is green and re-verified the acceptance testing around the adjusted conditional, merging!

```
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Update (118.41s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Add (125.06s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_Remove (141.67s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier (166.14s)
```

File: aws/resource_aws_rds_global_cluster.go (hunk 3 above)
Comment (mbfrahry) [reply aws/resource_aws_rds_global_cluster.go]: Ouch, that extra flag seems tricky but I'm totally onboard with seeing this come in later if it's requested.

**Comment by bflad** [comment]
This has been released in [version 1.53.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1530-december-20-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
